sockeye/data_io.py

```diff
@@ -1069,20 +1132,30 @@ def save(self, fname: str):
         """
         Saves the dataset to a binary .npy file.
         """
-        mx.nd.save(fname, self.source + self.target + self.label)
+        sources = self.source + self.target + self.label
+        if self.source_factors is not None:
+            for i in range(len(self.source_factors)):
+                sources += self.source_factors[i]
+        mx.nd.save(fname, sources)
 
     @staticmethod
-    def load(fname: str) -> 'ParallelDataSet':
+    def load(fname: str, num_source_factors: int = 0) -> 'ParallelDataSet':
         """
         Loads a dataset from a binary .npy file.
+        The first 1..num_source streams are the source and the next num_target are the target.
+        The last stream is the set of labels.
         """
         data = mx.nd.load(fname)
-        n = len(data) // 3
-        source = data[:n]
-        target = data[n:2 * n]
-        label = data[2 * n:]
+        num_parts = num_source_factors + 3
```

Comment:
If you move the factor data into source its cleaner and simplifies the loading logic a lot